src/Microsoft.Extensions.Configuration.AzureAppConfiguration/OfflineFileCache.cs

```diff
@@ -144,23 +103,25 @@ public string Import(AzureAppConfigurationOptions options)
                         }
                     }
 
-                    if ((data != null) && (dataHash != null) && (scopeHash != null))
+                    if ((data != null) && (scope != null) && (_scopeToken == scope))
                     {
-                        string newScopeHash = CryptoService.GetHash(Encoding.UTF8.GetBytes(_scopeToken), _options.SignKey);
-                        if (string.CompareOrdinal(scopeHash, newScopeHash) == 0)
-                        {
-                            string newDataHash = CryptoService.GetHash(Convert.FromBase64String(data), _options.SignKey);
-                            if (string.CompareOrdinal(dataHash, newDataHash) == 0)
-                            {
-                                return CryptoService.AESDecrypt(data, _options.Key, _options.IV);
-                            }
-                        }
+                        return _timeLimitedDataProtector.Unprotect(data);
                     }
                 }
                 catch (IOException ex) when (ex.HResult == ERROR_SHARING_VIOLATION)
                 {
                     Task.Delay(new Random().Next(delayRange)).ConfigureAwait(false).GetAwaiter().GetResult();
                 }
+                catch (CryptographicException)

```

Comment:
The expired data case is interesting. I imagine we should treat that as if the file doesn't exist at all rather than throwing.

Reply:
I wanted to handle `CryptographicException` in `AzureAppConfigurationProvider` where `Import` is called. Provider might need to decide whether to rethrow this exception with a good error message, or silently fail.

Reply:
> I wanted to handle CryptographicException in AzureAppConfigurationProvider

CryptographicException is quite generic. Can we distinguish the case where the data has expired? That's the only one I'm thinking about at the moment. The rest will most likely go unhandled.

Reply:
I think we should treat expired data case the same as the file not existing.

Reply:
I checked the HResult for different CryptographicExceptions. They always have the same HResult (-2146233087). Only the message is different. For example:
- Expired data = "The payload expired at 6/23/2021 11:22:04 PM +00:00."
- Tampered data = "The provided payload cannot be decrypted because it was not protected with this protection provider."
- Deleted encryption keys = "The provided payload cannot be decrypted because it was not protected with this protection provider."
- Some old invalid data (not sure how it was invalid) = "The payload was invalid."

Reply:
Perhaps we should persist the expiration date unencrypted so that we can avoid an exception in the expired data case.

Reply:
If we persist the expiration date unencrypted, someone can manually edit the expiration time to use the file forever. In that case, we will have to include something like a digital signature for the entire file to make sure that the expiration time hasn't been tampered with. That is an extra step that the time limited data protector saves.

Reply:
It doesn't change the fact that we use a time limited data protector. If the date gets extended we get a cryptographic exception when we try to decrypt the data. But if the user wants to cause a cryptographic exception all they need to do is modify the encrypted payload and they could do the same thing.

Reply:
In that case, its possible to get a false positive for expired data. Consider this:
Someone manually modified the expiration date in the cache file to a time in the past. They also tampered with the encrypted data. Now when we get a CryptographicException, we are going to deduce that this is because the file has expired, when in fact its the data that is corrupt.

I'm not sure why we need to differentiate between expired data vs other cryptographic exceptions. I agree the behavior in case of expired data and cache file not being present is the same. But isn't the end result same for all other cryptographic exceptions too - throw an exception and let users deal with it?

Reply:
Storing the expiration time in the file and using that for two purposes:
1. If expiration time has already passed, avoid CryptographicException by not decrypting the data.
2. If cache file doesn't contain expiration time, it could be an old cache file that the new provider will not be able to decrypt. Avoid another CryptographicException by not decrypting files which don't have expiration time.